### ISSUE TITLE
Virtual object bulk actions restart unexpectedly

### ISSUE THREAD

**Issue by andrewjbtw** [sul-dlss/argo#1728]
**Describe the bug**
I recently ran a virtual object bulk action job on a large CSV. In preparation for this job, the parent objects had been registered and accessioned to version 1. As a result of the job, successful virtual objects should have only been updated to version 2. But at some point the job began to restart itself from the top row, and by the time I made changes to stop it from repeatedly remaking virtual objects it had made up to 5 versions of some objects but had not updated others at all.

The full CSV can be found attached to this ticket: https://jirasul.stanford.edu/jira/browse/LEGACY-4586

Most of the parent objects from the CSV can be found in Argo with this search: https://argo.stanford.edu/catalog?f%5Btag_ssim%5D%5B%5D=Project+%3A+Stanford+Band+%28Series+3+Parents%29

You can see that some objects in that search now have as many as 6 versions, while most objects were not changed at all.

Note that the parent objects that are now on version 2 were not modified by the bulk action. In order to prevent the bulk action job from repeatedly running on the same objects over the weekend, I changed the rights to all parents to citation-only, which meant creating new versions for the rights update. 

**Timeline**

As best as I can tell, the bulk action did not start re-running the top rows until a number of hours had passed. Times below are approximate, but I think accurate to within 30 minutes:

***Friday***

2019-11-15 3:30 PM: started the bulk action job

2019-11-15 10:00 PM: checked on the items and the first few parents (maybe 10-15?) now had a version 2. Items looked correct. There were some SSL/network errors that I cleared on the workflow grid.

***Saturday***

2019-11-16 8:00 AM: checked again and saw that there was a set of parent objects with 3-5 versions. Checked on the "Opened" items and concluded that the bulk action was running on the top rows of the CSV again. Set parents to citation-only in attempt to prevent further repetition before we all came back to work Monday.

2019-11-18: Saw that the bulk action was still running, requested that we stop the job by killing the process. Even if it wasn't still creating objects, it was still checking their statuses to evaluate whether they could be merged.

**User Impact**
As long as a virtual object bulk action might run multiple times on the same object in a CSV, it is too risky to run it with CSVs with many rows, especially CSVs with long rows (many children). Virtual object bulk actions will need to be broken up into many parts.

**To Reproduce**
1. Submit a virtual object CSV with multiple rows and a large number of child druids on each row. It might require at least 10 rows and 1000 druids distributed across these rows to trigger this behavior.
2. Watch to see if it restarts.

**Expected behavior**
The bulk action should try to make a virtual object on each row in the CSV once before ever re-starting from the top.

**Additional context**
I believe that bulk actions are set to retry twice in some circumstances, but we may need to set the virtual object bulk action to never restart.

**Comment by ndushay** [comment]
Some explanations:

### What code is executed where for Virtual Object Creation:

- Argo:  it's a bulk action 
    - bulk actions are implemented with "delayed_job" to work async in the background
        - here are the settings used with delayed_job (from https://github.com/sul-dlss/argo/blob/master/config/initializers/delayed_job.rb
            - `max_run_time` 48 hours
            - `max_attempts` 3
    - https://github.com/sul-dlss/argo/blob/master/app/jobs/create_virtual_objects_job.rb
        - which calls https://github.com/sul-dlss/argo/blob/master/app/services/virtual_objects_creator.rb
            - which calls dor-services-app (see next leftmost bullet point)
            - it is set up to poll dor-services-app every `SECONDS_BETWEEN_REQUESTS` seconds (https://github.com/sul-dlss/argo/blob/master/app/services/virtual_objects_creator.rb#L5) to see if the entire csv of virtual objects has been completed.
- Dor-Services-App - it's a background job (async)
    - virtual object API endpoint uses "sidekiq" for async background processing.
        - we use the default settings in sidekiq (at this time). https://github.com/mperham/sidekiq/wiki/Advanced-Options
            - `retry` - defaults to true
    - https://github.com/sul-dlss/dor-services-app/blob/master/app/jobs/create_virtual_objects_job.rb
        - note that we currently only trap for certain exceptions ("rescue ActiveFedora::ObjectNotFoundError, Rubydora::FedoraInvalidRequest, Dor::Exception => e")
        - calls ConstituentService class:  https://github.com/sul-dlss/dor-services-app/blob/master/app/services/constituent_service.rb
            - talks to Fedora
            - opens and closes parent and each child version (hits workflows)
            - etc.
    -  (happy path) eventually background job finishes and is marked complete
- (happy path) argo polling finds "complete" for the job and marks it's own async process complete.

    
### What can be done to improve things?

(easy ones are marked "easy")

- dor-services-app traps more broadly for exceptions so it can mark the async job completed (with errors) if it gets borked (see https://github.com/sul-dlss/dor-services-app/blob/master/app/jobs/create_virtual_objects_job.rb#L18) - easy
- dor-services-app adds an "ensure" clause to this job so we definitely mark it completed with errors when it fails catestrophically. - easy
- dor-services-app allows for result codes beyond "'pending', 'processing', 'complete'"  (https://github.com/sul-dlss/dor-services-app/blob/master/db/migrate/20190917215521_create_background_job_results.rb#L7) so we can let argo know when job failed. (?) 
- increase the time between each argo poll to dor-services-app ("is it done yet?") - easy
- increase the max time in Argo for delayed_job to run - easy
- reduce the number of retries in Argo for virtual object delayed_job so we don't do inadvertent reruns (not sure if it could be done _only_ for virtual object jobs)
- argo creates a "webhook" sort of callback so dor-services-app can _push_ a notification when it's done processing, rather than having Argo do the polling. 
    - lots of polling may be putting unnec traffic on dor-services-app and network (currently set to every 3 seconds)
- switch Argo to use sidekiq instead of delayed_job for async processing.  This would afford us
    - a way to have a GUI to restart failures (see https://dor-services-prod.stanford.edu/queues/)

- have a way to know which virtual object merges in a large spreadsheet are completed and which are still to go
    - split it up at dor-services-app, per parent object, and do it in parallel?

### What could be researched further

- Sidekiq
   - what causes an automatic restart of a background job? (e.g. network outage?  app restart?)
   - are there settings we could leverage for more desirable outcomes?

- Delayed Job
   - what causes an automatic restart of a background job? (e.g. network outage?  app restart?)

- How to speed up the work on each object
    - can we make opening / closing versions faster?  Can we make updating rels-ext faster?  Can we make updating content-metadata faster?

- Why do we have so many network glitches?

- What are the failure patterns?
    - Fedora writes?
    - Knowing whether job is still running (so shouldn't be restarted)?  At dor-services-app level, or at argo level?
    - etc

**Comment by ndushay** [comment]
I will do the following as first responder, per Mike's comments:

- [ ] increase Argo's interval between polling requests ("is it done yet?") from 3 to 10 seconds
- [ ] have dor-services-app trap more broadly for errors when processing virt obj work
- [ ] add ensure clause to virt obj method in dor-services-app

**Comment by mjgiarlo** [comment]
@ndushay also worth considering, I think:

- [ ] do not retry virtobj jobs in Argo 

To do this, add this code:

```ruby
def max_attempts
  1
end
```

hereish: https://github.com/sul-dlss/argo/blob/master/app/jobs/create_virtual_objects_job.rb#L12

**Comment by ndushay** [comment]
okay, all 5 "easy" fixes have been merged and will be deployed on Monday.  We *think* this will result
- in fewer automatic restarts of virt obj merges (hopefully ... none?)
- failures coming back with info so you can hopefully see which parent objects errored, thus you can remove earlier parent object rows from a csv
- possibly fewer network hiccups to dor-services-app due to increasing poll interval

**Comment by mjgiarlo** [comment]
> okay, all 5 "easy" fixes have been merged and will be deployed on Monday. We _think_ this will result
> 
> * in fewer automatic restarts of virt obj merges (hopefully ... none?)
> * failures coming back with info so you can hopefully see which parent objects errored, thus you can remove earlier parent object rows from a csv
> * possibly fewer network hiccups to dor-services-app due to increasing poll interval

@andrewjbtw I'm checking to see if :point_up: has been helping things over the past couple weeks. Let us know if not!

**Comment by andrewjbtw** [comment]
As luck would have it, I'm getting ready to re-submit that earlier csv today. But this time not all at once.

**Comment by andrewjbtw** [comment]
The mysterious restarts are continuing. This time the job I'm running has only 10 lines in the CSV and the first few parent druids have been repeatedly run multiple times since last night.

Attached is the CSV for the currently running action (but with a .txt extension to satisfy github's filetype requirements) [PC0170_series3_combinator-1-10.txt](https://github.com/sul-dlss/argo/files/3933204/PC0170_series3_combinator-1-10.txt)

I think I'd better run further tests on stage and move through the problem CSV more or less one line at a time.

**Comment by mjgiarlo** [comment]
@andrewjbtw Boo, bummed to hear that. Doing further testing on stage makes sense.

**Comment by mjgiarlo** [comment]
OK, looking over Honeybadger alerts for the failed run from yesterday (12/5) around 7pm:

1. https://app.honeybadger.io/projects/50568/faults/56344007
  * For some reason, dor-services-app loses its connection to Postgres, which is the backend used for background results. This happens fifty minutes before the Argo job errors below, so I am not sure whether this is the main culprit or not. Given all the Redis connection problems we've been seeing in our SDR infrastructure, I suppose I am not surprised to see other connection problems occur. :grimacing: 
1. https://app.honeybadger.io/projects/49894/faults/56341350
  * This alert, primarily put there to help troubleshoot virtual object creation errors, is triggered when Argo receives virtual object creation results that includes errors. Notice that the first seven entries in the errors array are not JSON errors at all, but rather `nil`s, which are KNOWN to cause problems in the error parsing code. We could/should do further spelunking and analysis to figure out why/how these `nil`s got into the errors array.
1. https://app.honeybadger.io/projects/49894/faults/56344711
  * Because of the `nil`s in the errors array, this error is raised. We *could* ignore `nil`s in the errors array, but I fear that these `nil`s correspond with failed creation of virtual objects, so it is likely that doing this will misreport to the user what has happened with their batch of virtual objects.

**Comment by mjgiarlo** [comment]
On closer inspection, `nil`s in the errors array are what we get when virtual object creation SUCCEEDS. `:blink:`

**Comment by andrewjbtw** [comment]
The first seven parent druids on the CSV all got four runs. Parent druids 8, 9, and 10 all got one run, then got re-opened for a second run and are still in an "opened" state. It's unclear to me if the job is still active or not.

On the plus side, it looks like I can consider parent druids 1-7 "done" and take them off the remaining list.

**Comment by mjgiarlo** [comment]
OK, I found a definite bug in dor-services-app, which may be contributing to the weirdness on the Argo side: [PR](https://github.com/sul-dlss/dor-services-app/pull/513).

**Comment by mjgiarlo** [comment]
@andrewjbtw Now that the PR above is merged and deployed to stage and prod, I am eager to see whether it improves things. Submitting a batch known to contain both failures and successes would help us learn more.

**Comment by andrewjbtw** [comment]
Unfortunately, the restarts are still happening. I ran a couple of tests yesterday with lines from this CSV:

[first-four-lines.txt](https://github.com/sul-dlss/argo/files/4041555/first-four-lines.txt)

I ran the first test with only the first two lines from that file and the virtual object job was a success.

I ran the second test with all four lines and the job has been repeatedly re-starting since last night. There were two honeybadger errors in the middle of the night:

https://app.honeybadger.io/projects/49894/faults/58563736
https://app.honeybadger.io/projects/49894/faults/56027867

Looking at the timestamps of the [first druid on the list](https://argo-stage.stanford.edu/view/druid:cv948dz5616) the restarts happened before those errors, so it's still not clear why things are restarting. That druid started the job by creating version 18 and is now on version 25.

**Comment by justinlittman** [comment]
A few simple things to try to address `ActiveRecord::StatementInvalid: Mysql2::Error::ConnectionError: MySQL server has gone away`:
1. Set `reconnect: true` in `database.yml`.
2. Try `ActiveRecord::Base.verify_active_connections!` or `ActiveRecord::Base.clear_action_connections!` where appropriate in the job.

**Comment by justinlittman** [comment]
Setting `reconnect: true` didn't work.

I think we want to try something like [clear_active_connections!](https://apidock.com/rails/ActiveRecord/ConnectionAdapters/ConnectionHandler/clear_active_connections%21) around https://github.com/sul-dlss/argo/blob/master/app/jobs/create_virtual_objects_job.rb#L61

**Comment by peetucket** [comment]
> I think we want to try something like [clear_active_connections!](https://apidock.com/rails/ActiveRecord/ConnectionAdapters/ConnectionHandler/clear_active_connections%21) around [/app/jobs/create_virtual_objects_job.rb@`master`#L61](https://github.com/sul-dlss/argo/blob/master/app/jobs/create_virtual_objects_job.rb#L61) [](/sul-dlss/argo/blob/HEAD@{2020-01-22T17:18:22Z}/app/jobs/create_virtual_objects_job.rb#L61)

PR here to try this (waiting for stage to be available for testing): #1775

**Comment by ndushay** [comment]
@andrewjbtw is this finished and close-able (if  so, go head) or is there some piece of this remaining?

**Comment by andrewjbtw** [comment]
Confirmed this is working in prod. So far I've done a batch of 10 and a batch of 20, each with thousands of druids in total, and there hasn't been a problem.
